### ISSUE TITLE
Rename should report a conflict when renaming Deconstruct

### ISSUE THREAD

**Issue by SLaks** [dotnet/roslyn#16102]
**Version Used**: 2017 RC Refresh

_pedantic_ :)

```C#
class C {
	C() {
		var (a, b) = new C();
	}

	public void Deconstruct(out int a, out int b) => a = b = 1;
}
```

Renaming `Deconstruct` to anything else should report an unresolvable conflict on the deconstructing declaration.

For maximum awesomeness, it should resolve the conflict to
```C#
new C().Deconstruct2(out int a, out int b);
```

**Comment by SLaks** [comment]
Similarly, and worse, renaming to a better-matching `Deconstruct` will silently change the meaning of the callsite without reporting a conflict.

```C#
class Base { }
class C : Base {
	C() {
		var (a, b) = new C();
		a = b;
	}

	void Deconstruct2(out string a, out int b) { a = null; b = 0; }
}

static class Ex {
	public static void Deconstruct(this Base o, out int a, out int b) => a = b = 1;
}
```

Rename `Deconstruct2` to `Deconstruct`.

**Comment by jcouv** [comment]
This doesn't seem obvious to me. Such renaming could be intentional and useful. Maybe you do want to replace the method used for deconstruction.

**Comment by SLaks** [comment]
No; the prime directive of rename is to not change the semantic meaning at all.

**Comment by CyrusNajmabadi** [comment]
Renaming should be allowed.  but it should also notify when it will break something.

**Comment by SLaks** [comment]
Yes; exactly.  (or resolve the conflict)

**Comment by jcouv** [comment]
FYI @Pilchie I'll let IDE team prioritize this suggestion for Rename.

**Comment by Pilchie** [comment]
Note that fixing this probably means updating `SyntaxTreeInfo` to record whether a given file contains any deconstructions.

**Comment by jcouv** [comment]
Thanks for pointing this out @Pilchie 
If that helps, there is already an issue which will need indexing deconstructions https://github.com/dotnet/roslyn/issues/18963 (which I will try to implement in 15.6).

**Comment by jcouv** [comment]
This was fixed as part of implementing FindAllReferences for Deconstruct methods (PR https://github.com/dotnet/roslyn/pull/22934) in 15.6.

Here are the [tests](https://github.com/dotnet/roslyn/pull/22934/files#diff-155288bd1ba17a47966d80352348ba4c) that verify that a conflict warning is produced for renaming a Deconstruct method that is used in a deconstruction.

Closing this issue.
